src/librustdoc/passes/automatic_links.rs

```diff
@@ -0,0 +1,128 @@
+use super::{span_of_attrs, Pass};
+use crate::clean::*;
+use crate::core::DocContext;
+use crate::fold::DocFolder;
+use crate::html::markdown::opts;
+use core::ops::Range;
+use pulldown_cmark::{Event, LinkType, Parser, Tag};
+use regex::Regex;
+use rustc_errors::Applicability;
+use rustc_feature::UnstableFeatures;
+use rustc_session::lint;
+
+pub const CHECK_AUTOMATIC_LINKS: Pass = Pass {
+    name: "check-automatic-links",
+    run: check_automatic_links,
+    description: "detects URLS/email addresses that could be written using angle brackets",
+};
+
+const URL_REGEX: &str =
+    r"https?://(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)";
```

Comment:
Rather than hardcoding www, can you make this work for arbitrary subdomains?
```suggestion
    r"https?://([-a-zA-Z0-9@:%._\+~#=]{2,256}\.)+[a-zA-Z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)";
```
Please also add some tests, things like `https://blog.rust-lang.org` are fine.

Reply:
> `\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)`

What is this doing?

Reply:
Checking for url parameters I assume, like: `?search=12`